applications/ThresholdSignature.md

```diff
@@ -0,0 +1,300 @@
+
+```rust
+pub mod frost {
+    /// Generates the lagrange coefficient for the i'th participant.
+    ///
+    /// Implements [`derive_interpolating_value()`] from the spec: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-15.html#name-polynomials
+    fn derive_interpolating_value(signer_id: &Identifier, signing_package: &SigningPackage) -> Result<Scalar, Error>;
+
+    /// Encodes the list of group signing commitments.
+    ///
+    /// Implements [`encode_group_commitment_list()`] from the spec: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-15.html#name-list-operations
+    fn encode_group_commitment_list(signing_commitments: &BTreeMap<Identifier, SigningCommitments>) -> Vec<u8>;
+
+    /// Extracts identifiers from a commitment list.
+    ///
+    /// Implements [`participants_from_commitment_list()`] from the spec: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-15.html#section-4.3-3
+    fn participants_from_commitment_list(commitment_list: &[(Identifier, NonceCommitment, NonceCommitment)]) -> Vec<Identifier>;
+
+    /// Extracts a binding factor from a list of binding factors.
+    ///
+    /// Implements [`binding_factor_for_participant()`] from the spec: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-15.html#section-4.3-5
+    fn binding_factor_for_participant(binding_factor_list: &BindingFactorList, identifier: Identifier) -> Result<BindingFactor, Error>;
+
+    /// Computes binding factors based on the participant commitment list, message to be signed, and group public key.
+    ///
+    /// Implements [`compute_binding_factor`] from the spec: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-15.html#section-4.4
+    pub(crate) fn compute_binding_factor(signing_package: &SigningPackage, verifying_key: &VerifyingKey, additional_prefix: &[u8]) -> BindingFactorList;
+
+    /// Generates the group commitment which is published as part of the joint Schnorr signature.
+    ///
+    /// Implements [`compute_group_commitment`] from the spec: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-15.html#section-4.5
+    fn compute_group_commitment(signing_package: &SigningPackage, binding_factor_list: &BindingFactorList) -> Result<GroupCommitment, Error>;
+
+    /// Generates the challenge as is required for Schnorr signatures.
+    ///
+    /// Implements [`compute_challenge`] from the spec: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-15.html#name-signature-challenge-computa
+    fn compute_challenge(R: &GroupCommitment, verifying_key: &VerifyingKey, msg: &[u8]) -> Challenge;
+
+    /// Generates the signing nonces and commitments to be used in the signing operation.
+    ///
+    /// Implements [`commit`] from the spec: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-15.html#name-round-one-commitment
+    pub fn commit(secret: &SigningShare) -> (SigningNonces, SigningCommitments);
```

Comment:
This could take a `&impl Transcript` too, which'll add entropy.  It's only immutable though, so no transcript modifications.